src/v/cluster/tm_stm.cc

```diff
@@ -107,32 +124,71 @@ tm_stm::update_tx(tm_transaction tx, model::term_id term) {
         co_return tm_stm::op_status::unknown;
     }
 
-    auto ptx = _tx_table.find(tx.id);
-    if (ptx == _tx_table.end()) {
+    auto tx_opt = get_tx(tx.id);
+    if (!tx_opt.has_value()) {
         co_return tm_stm::op_status::conflict;
     }
-    co_return ptx->second;
+    co_return tx_opt.value();
 }
 
 ss::future<checked<tm_transaction, tm_stm::op_status>>
-tm_stm::try_change_status(
-  kafka::transactional_id tx_id, tm_transaction::tx_status status) {
-    if (!_c->is_leader()) {
-        co_return tm_stm::op_status::not_leader;
+tm_stm::mark_tx_preparing(kafka::transactional_id tx_id) {
+    auto ptx = _mem_txes.find(tx_id);
+    if (ptx == _mem_txes.end()) {
+        co_return tm_stm::op_status::not_found;
+    }
+    auto tx = ptx->second;
+    if (tx.status != tm_transaction::tx_status::ongoing) {
+        co_return tm_stm::op_status::conflict;
     }
+    tx.status = cluster::tm_transaction::tx_status::preparing;
+    tx.last_update_ts = clock_type::now();
+    co_return co_await update_tx(tx, tx.etag);
+}
 
-    auto is_ready = co_await sync(_sync_timeout);
-    if (!is_ready) {
-        co_return tm_stm::op_status::unknown;
+ss::future<checked<tm_transaction, tm_stm::op_status>>
+tm_stm::mark_tx_aborting(kafka::transactional_id tx_id) {
```

Comment:
it seems like this commit could be broken down into several commits, such as introducing these specializations of `try_change_status`, return type change of `get_expired_txs`, etc... if not I can try to review but it seems like several things going on here at once.

Reply:
I'll break it, good observation, sometimes it's very tempting to change a couple of things along the way

Reply:
done